### DISCUSSION TITLE
Are custom keys not allowed for useFieldArray?

### DISCUSSION THREAD

**Discussion by zsidnam** [react-hook-form/react-hook-form#5479] [Q&A] [answered]
It looks like RHF generates its own unique identifiers that are used as keys in a field array. Is there a reason that custom keys cannot be used?

It appears that if a record in the field array has a property with the same name as RHF keyName, it will overwrite the RHF key and cause issues with controlled components, even though the non-RHF keys are unique. I modified the CodePen of the `useFieldArray` example from the RHF docs to show an example of what happens when I provide a custom key and use with controlled components: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-controlled-component-key-example-pmbwl

I don't think there is necessarily an issue here, but I think the docs could be more clear that RHF is generating its own keys and you need to make sure that the records in the field array don't have conflicting fields. (Looks like there is a `keyName` option for `useFieldArray` to avoid this). I discovered this because I was converting an existing form that used Formik + MUI over to RHF, so I was using controlled components and had kept the existing keys that the Formik `FieldArray` component was using.

Please let me know if I am misunderstanding, though! I spent enough time trying to figure out what was wrong with my form  that I wanted to post here and possibly save someone else the trouble.

Btw, this library is amazing!!! I started using it because I had a form with a ton of fields and the performance of Formik was borderline unusable. Switched over to RHF and works extremely well now!

**Comment by bluebill1049** [comment, marked as answer]
hey @zsidnam thanks for the feedback. It's a valid call out, I will update the doc on this piece of your information. Yes, you can use your own key as long as you keep those keys consistent through the render. I will add more details to the doc.

**Reply by zsidnam** [reply]
@bluebill1049 thanks so much for the quick response! Excellent job maintaining this library 👏 

If that is the case, unless I'm missing something, I think there might be a bug for controlled components when using a custom key as shown in the CodePen I linked. Regardless, I definitely think some clarification in the docs would be really helpful, so thanks for looking into that!

If there's anything I can do help with this, please let me know!

**Reply by bluebill1049** [reply]
Can you change ur key into a hash or something more complex than just 1 and 2?

**Reply by zsidnam** [reply]
I was using uuid to generate keys and got the same issue. I have updated the CodeSandbox above to show that.

**Reply by bluebill1049** [reply]
It's working for me. when i delete the first item. id need to be unique and not changing during the render as well.

**Reply by zsidnam** [reply]
If you follow these steps, you'll see what I mean:
1. Delete the first record (Bill Luo) -> everything works as expected, you are left with only Jack Sparrow
2. Hit Reset -> the first record is added back in, but now the second record has the properties of the first record copied onto it (both of them now say Bill Luo)
3. Hit Reset again -> return to original form state (Bill Luo and Jack Sparrow)

Something weird is happening on that first reset click. I verified that it is not just a render issue; if you hit submit after step 2, the data will show that both records are Bill Luo. This only happens for controlled fields.

**Reply by bluebill1049** [reply]
provide `defaultValues` to the reset should resolve this issue.

**Reply by zsidnam** [reply]
It looks like it is intended that calling reset() with no arguments should work since default values are passed into the useForm hook, but I can close this out if that is not supported. I don't have to pass in default values to reset if I let RHF generate the id field and it works as expected.

Thanks for your help!
